File: src/com/evo/NEAT/Genome.java
package com.evo.NEAT;

import com.evo.NEAT.com.evo.NEAT.config.NEAT_Config;

import javax.management.RuntimeErrorException;
import java.io.*;
import java.util.*;

/**
 * Created by vishnughosh on 28/02/17.
 */
public class Genome implements Comparable {
    private static Random rand = new Random();
    private float fitness;                                          // Global Percentile Rank (higher the better)
    private float points;
    private ArrayList<ConnectionGene> connectionGeneList = new ArrayList<>();           // DNA- MAin archive of gene information
    private TreeMap<Integer, NodeGene> nodes = new TreeMap<>();                          // Generated while performing network operation
    private float adjustedFitness;                                      // For number of child to breed in species

    private HashMap<MutationKeys, Float> mutationRates = new HashMap<>();

    private enum MutationKeys {
        STEPS,
        PERTURB_CHANCE,
        WEIGHT_CHANCE,
        WEIGHT_MUTATION_CHANCE,
        NODE_MUTATION_CHANCE,
        CONNECTION_MUTATION_CHANCE,
        BIAS_CONNECTION_MUTATION_CHANCE,
        DISABLE_MUTATION_CHANCE,
        ENABLE_MUTATION_CHANCE
    }
    /*    private class MutationRates{
            float STEPS;
            float PERTURB_CHANCE;
            float WEIGHT_CHANCE;
            float WEIGHT_MUTATION_CHANCE;
            float NODE_MUTATION_CHANCE;
            float CONNECTION_MUTATION_CHANCE;
            float BIAS_CONNECTION_MUTATION_CHANCE;
            float DISABLE_MUTATION_CHANCE;
            float ENABLE_MUTATION_CHANCE;

             MutationRates() {
                this.STEPS = NEAT_Config.STEPS;
                this.PERTURB_CHANCE = NEAT_Config.PERTURB_CHANCE;
                this.WEIGHT_CHANCE = NEAT_Config.WEIGHT_CHANCE;
                this.WEIGHT_MUTATION_CHANCE = NEAT_Config.WEIGHT_MUTATION_CHANCE;
                this.NODE_MUTATION_CHANCE = NEAT_Config.NODE_MUTATION_CHANCE;
                this.CONNECTION_MUTATION_CHANCE = NEAT_Config.CONNECTION_MUTATION_CHANCE;
                this.BIAS_CONNECTION_MUTATION_CHANCE = NEAT_Config.BIAS_CONNECTION_MUTATION_CHANCE;
                this.DISABLE_MUTATION_CHANCE = NEAT_Config.DISABLE_MUTATION_CHANCE;
                this.ENABLE_MUTATION_CHANCE = NEAT_Config.ENABLE_MUTATION_CHANCE;
            }
        }*/
    public Genome(){

        this.mutationRates.put(MutationKeys.STEPS, NEAT_Config.STEPS);
        this.mutationRates.put(MutationKeys.PERTURB_CHANCE, NEAT_Config.PERTURB_CHANCE);
        this.mutationRates.put(MutationKeys.WEIGHT_CHANCE,NEAT_Config.WEIGHT_CHANCE);
        this.mutationRates.put(MutationKeys.WEIGHT_MUTATION_CHANCE, NEAT_Config.WEIGHT_MUTATION_CHANCE);
        this.mutationRates.put(MutationKeys.NODE_MUTATION_CHANCE , NEAT_Config.NODE_MUTATION_CHANCE);
        this.mutationRates.put(MutationKeys.CONNECTION_MUTATION_CHANCE , NEAT_Config.CONNECTION_MUTATION_CHANCE);
        this.mutationRates.put(MutationKeys.BIAS_CONNECTION_MUTATION_CHANCE , NEAT_Config.BIAS_CONNECTION_MUTATION_CHANCE);
        this.mutationRates.put(MutationKeys.DISABLE_MUTATION_CHANCE , NEAT_Config.DISABLE_MUTATION_CHANCE);
        this.mutationRates.put(MutationKeys.ENABLE_MUTATION_CHANCE , NEAT_Config.ENABLE_MUTATION_CHANCE);
    }

    public Genome(Genome child) {

        for (ConnectionGene c:child.connectionGeneList){
            this.connectionGeneList.add(new ConnectionGene(c));
        }

        this.fitness = child.fitness;
        this.adjustedFitness = child.adjustedFitness;

        this.mutationRates = (HashMap<MutationKeys, Float>) child.mutationRates.clone();

    }


    public float getFitness() {
        return fitness;
    }

    public void setFitness(float fitness) {
        this.fitness = fitness;
    }

    // Can remove below setter-getter after testing
    public ArrayList<ConnectionGene> getConnectionGeneList() {
        return connectionGeneList;
    }

    public void setConnectionGeneList(ArrayList<ConnectionGene> connectionGeneList) {
        this.connectionGeneList = connectionGeneList;
    }

    public static Genome crossOver(Genome parent1, Genome parent2){
        if(parent1.fitness < parent2.fitness){
            Genome temp = parent1;
            parent1 = parent2;
            parent2 = temp;
        }

        Genome child = new Genome();
        TreeMap<Integer, ConnectionGene> geneMap1 = new TreeMap<>();
        TreeMap<Integer, ConnectionGene> geneMap2 = new TreeMap<>();

        for(ConnectionGene con: parent1.connectionGeneList){
            assert  !geneMap1.containsKey(con.getInnovation());             //TODO Remove for better performance
            geneMap1.put(con.getInnovation(), con);
        }

        for(ConnectionGene con: parent2.connectionGeneList){
            assert  !geneMap2.containsKey(con.getInnovation());             //TODO Remove for better performance
            geneMap2.put(con.getInnovation(), con);
        }

        Set<Integer> innovationP1 = geneMap1.keySet();
        Set<Integer> innovationP2 = geneMap2.keySet();

        Set<Integer> allInnovations = new HashSet<Integer>(innovationP1);
        allInnovations.addAll(innovationP2);

        for(int key : allInnovations){
            ConnectionGene trait;

            if(geneMap1.containsKey(key) && geneMap2.containsKey(key)){
                if(rand.nextBoolean()){
                    trait = new ConnectionGene(geneMap1.get(key));
                }else {
                    trait = new ConnectionGene(geneMap2.get(key));
                }

                if((geneMap1.get(key).isEnabled()!=geneMap2.get(key).isEnabled())){
                    if( (rand.nextFloat()<0.75f ))
                        trait.setEnabled(false);
                    else
                        trait.setEnabled(true);
                }

            }else if(parent1.getFitness()==parent2.getFitness()){               // disjoint or excess and equal fitness
                if(geneMap1.containsKey(key))
                    trait = geneMap1.get(key);
                else
                    trait = geneMap2.get(key);

                if(rand.nextBoolean()){
                    continue;
                }

            }else
                trait = geneMap1.get(key);


            child.connectionGeneList.add(trait);
        }


        return child;

    }


    public static boolean isSameSpecies(Genome g1, Genome g2){
        TreeMap<Integer, ConnectionGene> geneMap1 = new TreeMap<>();
        TreeMap<Integer, ConnectionGene> geneMap2 = new TreeMap<>();

        int matching = 0;
        int disjoint = 0;
        int excess = 0;
        float weight = 0;
        int lowMaxInnovation;
        float delta = 0;

        for(ConnectionGene con: g1.connectionGeneList) {
            assert  !geneMap1.containsKey(con.getInnovation());             //TODO Remove for better performance
            geneMap1.put(con.getInnovation(), con);
        }

        for(ConnectionGene con: g2.connectionGeneList) {
            assert  !geneMap2.containsKey(con.getInnovation());             //TODO Remove for better performance
            geneMap2.put(con.getInnovation(), con);
        }
        if(geneMap1.isEmpty() || geneMap2.isEmpty())
            lowMaxInnovation = 0;
        else
            lowMaxInnovation = Math.min(geneMap1.lastKey(),geneMap2.lastKey());

        Set<Integer> innovationP1 = geneMap1.keySet();
        Set<Integer> innovationP2 = geneMap2.keySet();

        Set<Integer> allInnovations = new HashSet<Integer>(innovationP1);
        allInnovations.addAll(innovationP2);

        for(int key : allInnovations){

            if(geneMap1.containsKey(key) && geneMap2.containsKey(key)){
                matching ++;
                weight += Math.abs(geneMap1.get(key).getWeight() - geneMap2.get(key).getWeight());
            }else {
                if(key < lowMaxInnovation){
                    disjoint++;
                }else{
                    excess++;
                }
            }

        }

        //System.out.println("matching : "+matching + "\ndisjoint : "+ disjoint + "\nExcess : "+ excess +"\nWeight : "+ weight);

        int N = matching+disjoint+excess ;

        if(N>0)
            delta = (NEAT_Config.EXCESS_COEFFICENT * excess + NEAT_Config.DISJOINT_COEFFICENT * disjoint) / N + (NEAT_Config.WEIGHT_COEFFICENT * weight) / matching;

        return delta < NEAT_Config.COMPATIBILITY_THRESHOLD;

    }

    private void generateNetwork() {

        nodes.clear();
        //  Input layer
        for (int i = 0; i < NEAT_Config.INPUTS; i++) {
            nodes.put(i, new NodeGene(0));                    //Inputs
        }
        nodes.put(NEAT_Config.INPUTS, new NodeGene(1));        // Bias

        //output layer
        for (int i = NEAT_Config.INPUTS + NEAT_Config.HIDDEN_NODES; i < NEAT_Config.INPUTS + NEAT_Config.HIDDEN_NODES + NEAT_Config.OUTPUTS; i++) {
            nodes.put(i, new NodeGene(0));
        }

        // hidden layer
        for (ConnectionGene con : connectionGeneList) {
            if (!nodes.containsKey(con.getInto()))
                nodes.put(con.getInto(), new NodeGene(0));
            if (!nodes.containsKey(con.getOut()))
                nodes.put(con.getOut(), new NodeGene(0));
            nodes.get(con.getOut()).getIncomingCon().add(con);
        }


    }

    public float[] evaluateNetwork(float[] inputs) {
        float output[] = new float[NEAT_Config.OUTPUTS];
        generateNetwork();

        for (int i = 0; i < NEAT_Config.INPUTS; i++) {
            nodes.get(i).setValue(inputs[i]);
        }

        for (Map.Entry<Integer, NodeGene> mapEntry : nodes.entrySet()) {
            float sum = 0;
            int key = mapEntry.getKey();
            NodeGene node = mapEntry.getValue();

            if (key > NEAT_Config.INPUTS) {
                for (ConnectionGene conn : node.getIncomingCon()) {
                    if (conn.isEnabled()) {
                        sum += nodes.get(conn.getInto()).getValue() * conn.getWeight();
                    }
                }
                node.setValue(sigmoid(sum));
            }
        }

        for (int i = 0; i < NEAT_Config.OUTPUTS; i++) {
            output[i] = nodes.get(NEAT_Config.INPUTS + NEAT_Config.HIDDEN_NODES + i).getValue();
        }
        return output;
    }

    private float sigmoid(float x) {
        // TODO Auto-generated method stub
        return (float) (1 / (1 + Math.exp(-4.9 * x)));
    }

    // Mutations

    public void Mutate() {
        // Mutate mutation rates
        for (Map.Entry<MutationKeys, Float> entry : mutationRates.entrySet()) {
            if(rand.nextBoolean())
                mutationRates.put(entry.getKey(), 0.95f * entry.getValue() );
            else
                mutationRates.put(entry.getKey(), 1.05263f * entry.getValue() );
        }


        if (rand.nextFloat() <= mutationRates.get(MutationKeys.WEIGHT_MUTATION_CHANCE))
            mutateWeight();
        if (rand.nextFloat() <= mutationRates.get(MutationKeys.CONNECTION_MUTATION_CHANCE))
            mutateAddConnection(false);
        if (rand.nextFloat() <= mutationRates.get(MutationKeys.BIAS_CONNECTION_MUTATION_CHANCE))
            mutateAddConnection(true);
        if (rand.nextFloat() <= mutationRates.get(MutationKeys.NODE_MUTATION_CHANCE))
            mutateAddNode();
        if (rand.nextFloat() <= mutationRates.get(MutationKeys.DISABLE_MUTATION_CHANCE))
            disableMutate();
        if (rand.nextFloat() <= mutationRates.get(MutationKeys.ENABLE_MUTATION_CHANCE))
            enableMutate();
    }

    void mutateWeight() {

        for (ConnectionGene c : connectionGeneList) {
            if (rand.nextFloat() < NEAT_Config.WEIGHT_CHANCE) {
                if (rand.nextFloat() < NEAT_Config.PERTURB_CHANCE)
                    c.setWeight(c.getWeight() + (2 * rand.nextFloat() - 1) * NEAT_Config.STEPS);
                else c.setWeight(4 * rand.nextFloat() - 2);
            }
        }
    }

    void mutateAddConnection(boolean forceBais) {
        generateNetwork();
        int i = 0;
        int j = 0;
        int random2 = rand.nextInt(nodes.size() - NEAT_Config.INPUTS - 1) + NEAT_Config.INPUTS + 1;
        int random1 = rand.nextInt(nodes.size());
        if(forceBais)
            random1 = NEAT_Config.INPUTS;
        int node1 = -1;
        int node2 = -1;

        for (int k : nodes.keySet()) {
            if (random1 == i) {
                node1 = k;
                break;
            }
            i++;
        }

        for (int k : nodes.keySet()) {
            if (random2 == j) {
                node2 = k;
                break;
            }
            j++;
        }
//	System.out.println("random1 = "+random1 +" random2 = "+random2);
//	System.out.println("Node1 = "+node1 +" node 2 = "+node2);


        if (node1 >= node2)
            return;

        for (ConnectionGene con : nodes.get(node2).getIncomingCon()) {
            if (con.getInto() == node1)
                return;
        }

        if (node1 < 0 || node2 < 0)
            throw new RuntimeErrorException(null);          // TODO Pool.newInnovation(node1, node2)
        connectionGeneList.add(new ConnectionGene(node1, node2, InnovationCounter.newInnovation(), 4 * rand.nextFloat() - 2, true));                // Add innovation and weight

    }

    void mutateAddNode() {
        generateNetwork();
        if (connectionGeneList.size() > 0) {
            int timeoutCount = 0;
            ConnectionGene randomCon = connectionGeneList.get(rand.nextInt(connectionGeneList.size()));
            while (!randomCon.isEnabled()) {
                randomCon = connectionGeneList.get(rand.nextInt(connectionGeneList.size()));
                timeoutCount++;
                if (timeoutCount > NEAT_Config.HIDDEN_NODES)
                    return;
            }
            int nextNode = nodes.size() - NEAT_Config.OUTPUTS;
            randomCon.setEnabled(false);
            connectionGeneList.add(new ConnectionGene(randomCon.getInto(), nextNode, InnovationCounter.newInnovation(), 1, true));        // Add innovation and weight
            connectionGeneList.add(new ConnectionGene(nextNode, randomCon.getOut(), InnovationCounter.newInnovation(), randomCon.getWeight(), true));
        }
    }
    void disableMutate() {
        //generateNetwork();                // remove laters
        if (connectionGeneList.size() > 0) {
            ConnectionGene randomCon = connectionGeneList.get(rand.nextInt(connectionGeneList.size()));
            randomCon.setEnabled(false);
        }
    }


    void enableMutate() {
        //generateNetwork();                // remove laters
        if (connectionGeneList.size() > 0) {
            ConnectionGene randomCon = connectionGeneList.get(rand.nextInt(connectionGeneList.size()));
            randomCon.setEnabled(true);
        }
    }

    @Override
    public int compareTo(Object o) {
        Genome g = (Genome)o;
        if (fitness==g.fitness)
            return 0;
        else if(fitness >g.fitness)
            return 1;
        else
            return -1;
    }

    @Override
    public String toString() {
        return "Genome{" +
                "fitness=" + fitness +
                ", connectionGeneList=" + connectionGeneList +
                ", nodeGenes=" + nodes +
                '}';
    }

    public void setAdjustedFitness(float adjustedFitness) {
        this.adjustedFitness = adjustedFitness;
    }

    public float getAdjustedFitness() {
        return adjustedFitness;
    }

    public float getPoints() {
        return points;
    }

    public void setPoints(float points) {
        this.points = points;
    }

    public void writeTofile(){
        BufferedWriter bw = null;
        FileWriter fw = null;
        StringBuilder builder = new StringBuilder();
        for (ConnectionGene conn: connectionGeneList) {
            builder.append(conn.toString()+"\n");
        }
        try {


            fw = new FileWriter("Genome.txt");
            bw = new BufferedWriter(fw);
            bw.write(builder.toString());

            System.out.println("Done");

        } catch (IOException e) {

            e.printStackTrace();

        } finally {

            try {

                if (bw != null)
                    bw.close();

                if (fw != null)
                    fw.close();

            } catch (IOException ex) {

                ex.printStackTrace();

            }

        }

    }

}

File: src/main/java/com/evo/NEAT/Genome.kt
@file:Suppress("NOTHING_TO_INLINE")

package com.evo.NEAT

import com.evo.NEAT.Genome.Companion.INDEXABLE
import com.evo.NEAT.Genome.Companion.sim
import com.evo.NEAT.com.evo.NEAT.ActivationFunction
import com.evo.NEAT.config.NEAT_Config
import com.evo.NEAT.config.Sim
import javolution.util.FastSet
import javolution.util.FastSortedMap
import javolution.util.FastSortedTable
import javolution.util.FastTable
import javolution.util.function.Equality
import java.io.BufferedWriter
import java.io.FileWriter
import java.io.IOException
import java.util.*
import java.util.concurrent.atomic.AtomicInteger
import javax.management.RuntimeErrorException
import kotlin.math.max
import kotlin.math.min
import kotlin.random.Random

/**
 * Created by vishnughosh on 28/02/17.
 */
/*@Serializable*/
class Genome(
    /*@Serializable*/
    val nodes: FastSortedTable<NodeGene> = FastSortedTable<NodeGene>(object : Equality<NodeGene> {
        override fun compare(left: NodeGene, right: NodeGene): Int = left.key.compareTo(right.key)
        override fun hashCodeOf(it: NodeGene) = it.key
        override fun areEqual(left: NodeGene?, right: NodeGene?) = left?.key == right?.key
    }),
    /*@Serializable*/
    val mutationRates: MutableMap<MutationKeys, Double> = EnumMap(MutationKeys.mutationRates),
    /** Global Percentile Rank (higher the better)*/
    var fitness: Double = 0.0,
    var points: Double = 0.0,
    val connections: FastTable<ConnectionGene> = FastTable(),
    var adjustedFitness: Double = 0.0,
) : Comparable<Genome> {

//    mutationRates = MutationKeys.mutationRates

    constructor(parent: Genome) : this() {
        parent.connections.mapTo(connections) { ConnectionGene(it) }
        parent.nodes.mapTo(this.nodes) {
            NodeGene(it.takeUnless { (it.key == sim.INPUTS) }?: INPUTBIAS)
        }
        fitness = parent.fitness
        adjustedFitness = parent.adjustedFitness
        mutationRates.putAll(parent.mutationRates)
        points = parent.points
    }


    /* var color: ActivationFunction = ActivationFunction.values().random()*/

    fun generateNetwork() {//TODO: dirty flag

        if (nodes.isEmpty()) {
            nodes
                .apply { for (i in 0 until sim.INPUTS) add(NodeGene(i)) }
                .apply { add(INPUTBIAS) }
                .apply {
                    for (i in INDEXABLE until INDEXABLE + sim.OUTPUTS) add(NodeGene(i).also {
                        it.activationFunction = ActivationFunction.Linear
                    } /* extra supervision  */)
                }
        }

        // hidden layer
        connections.forEach { con ->
            if (!nodes.any { it.key == con.keyInto }) nodes.addLast(NodeGene(con.keyInto))
            if (!nodes.any { it.key == con.keyOut }) nodes.addLast(
                NodeGene(con.keyOut, con.weight, FastTable<ConnectionGene>().also { it += ConnectionGene(con) }))
        }
        assert(true)
    }


    fun evaluateNetwork(inputs: DoubleArray): DoubleArray {
        generateNetwork()
        for (i in 0 until sim.INPUTS) (+nodes)(i).impulse = inputs[i]
        nodes.filtered {
            it.key >= sim.INPUTS
        }.forEach { node ->
            val toDownStream = node.incomingCon.filter {
                it.isEnabled
            }.map { (inKey, _, _, weight, _): ConnectionGene ->
                val upstreamNode = (+nodes)(inKey)
                val impulse = upstreamNode.impulse
                val res = impulse * weight
                res
            }
            var final: Double
            toDownStream.sum().let { sum ->
                node.run {
                    final = activationFunction(sum)
                    impulse = final
                }
            }
            assert(true)
        }

        return DoubleArray(sim.OUTPUTS) { i ->
            (-nodes)(INDEXABLE + i).impulse
        }
    }

/*   private fun doColor(x: Double): Double {
       return color.apply(x.toDouble()).toDouble()
   }*/

    // Mutations
    fun Mutate() {
        // Mutate mutation rates
        for ((key, value) in mutationRates)
            if (rand.nextBoolean()) mutationRates[key] = 0.95 * value else mutationRates[key] =
                1.05263f * value
        for (it in MutationKeys.WEIGHT_MUTATION_CHANCE.ordinal..MutationKeys.ENABLE_MUTATION_CHANCE.ordinal) {
            MutationKeys.values()[it].let { mk ->
                if (this.mutationRates[mk]!!.toDouble() > rand.nextDouble()) mk.fn(this)
            }
        }
    }

    fun mutateWeight() {
        for (c in connections) if (rand.nextDouble() < NEAT_Config.WEIGHT_CHANCE)
            if (rand.nextDouble() < NEAT_Config.PERTURB_CHANCE) c.weight =
                c.weight + (rand.nextDouble(-1.0, (1.0 + Double.MIN_VALUE))) * NEAT_Config.STEPS else c.weight =
                rand.nextDouble(-2.0, 2.0)
    }


    /**
     * tmfm happens here.
     */
    fun mutateAddConnection(forceBais: Boolean) {
        generateNetwork()
        var i = 0
        var j = 0
        val random2 = rand.nextInt(nodes.size - sim.INPUTS - 1) + sim.INPUTS + 1
        var random1 = rand.nextInt(nodes.size)
        if (forceBais) random1 = sim.INPUTS
        var node1 = -1
        var node2 = -1
        for (k in nodes.mapped(NodeGene::key)) {
            if (random1 == i) {
                node1 = k
                break
            }
            i++
        }
        for (k in nodes.mapped(NodeGene::key)) {
            if (random2 == j) {
                node2 = k
                break
            }
            j++
        }
        //	System.out.println("random1 = "+random1 +" random2 = "+random2);
//	System.out.println("Node1 = "+node1 +" node 2 = "+node2);
        if (node1 >= node2) return
        for (con in (-nodes)(node2)!!.incomingCon) {
            if (con.keyInto == node1) return
        }
        if (node1 < 0 || node2 < 0) throw RuntimeErrorException(null) // TODO Pool.newInnovation(node1, node2)
        connections.add(ConnectionGene(node1,
            node2,
            innovation.getAndIncrement(),
            rand.nextDouble(),
            true)) // Add innovation and weight
    }

    fun mutateAddNode() {
        generateNetwork()
        if (connections.size > 0) {
            var timeoutCount = 0
            var randomCon = connections[rand.nextInt(connections.size)]
            while (!randomCon.isEnabled) {
                randomCon = connections[rand.nextInt(connections.size)]
                timeoutCount++
                if (timeoutCount > sim.HIDDEN_NODES) return
            }
            val nextNode = nodes.size - sim.OUTPUTS
            randomCon.isEnabled = false
            connections.add(
                ConnectionGene(
                    randomCon.keyInto,
                    nextNode,
                    /**
                     * identifying string
                     */
                    innovation.getAndIncrement(),
                    1.0,
                    true
                )
            ) // Add innovation and weight
            connections.add(
                ConnectionGene(
                    nextNode,
                    randomCon.keyOut,
                    innovation.getAndIncrement(),
                    randomCon.weight,
                    true
                )
            )
        }
    }

    fun disableMutate() {
        //generateNetwork();                // remove laters
        if (connections.size > 0) {
            val randomCon = connections[rand.nextInt(connections.size)]
            randomCon.isEnabled = false
        }
    }

    fun enableMutate() {
//        generateNetwork();                // remove laters
        if (connections.size > 0) {
            val randomCon = connections[rand.nextInt(connections.size)]
            randomCon.isEnabled = true
        }
    }

    override fun compareTo(other: Genome): Int = fitness.compareTo(other.fitness)


    fun writeTofile() {
        var bw: BufferedWriter? = null
        var fw: FileWriter? = null
        val builder = StringBuilder()
        for (conn in connections) {
            builder.append(
                """
    $conn
    
    """.trimIndent()
            )
        }
        try {
            fw = FileWriter("Genome.txt")
            bw = BufferedWriter(fw)
            bw.write(builder.toString())
            println("Done")
        } catch (e: IOException) {
            e.printStackTrace()
        } finally {
            try {
                bw?.close()
                fw?.close()
            } catch (ex: IOException) {
                ex.printStackTrace()
            }
        }
    }


    companion object {
        val rand = Random
        val innovation = AtomicInteger(0)

        val INPUTBIAS by lazy {
            object : NodeGene(sim.INPUTS, 1.0, FastTable(), ActivationFunction.Linear) {
                override val key get() = Genome.sim.INPUTS
                override var impulse
                    get() = 1.0
                    set(value) {}
                override var activationFunction
                    get() = ActivationFunction.Linear
                    set(value) {}
            }
        }

        @JvmStatic
        fun crossOver(p1: Genome, p2: Genome): Genome {
            var parent1 = p1
            var parent2 = p2

            if (parent1.fitness < parent2.fitness) {
                val temp = parent1
                parent1 = parent2
                parent2 = temp
            }
            val child = Genome()
            val geneMap1 = FastSortedMap<Int, ConnectionGene>()
            val geneMap2 = FastSortedMap<Int, ConnectionGene>()
            for (con in parent1.connections) geneMap1[con.innovationKey] = con
            for (con in parent2.connections) geneMap2[con.innovationKey] = con
            val innovationP1: FastSet<Int> = geneMap1.keys
            val innovationP2: FastSet<Int> = geneMap2.keys
            val allInnovations = FastSet<Int>()
            allInnovations.addAll(innovationP1)
            allInnovations.addAll(innovationP2)
            for (key in allInnovations) {
                var trait: ConnectionGene?
                if (geneMap1.containsKey(key) && geneMap2.containsKey(key)) {
                    trait = ConnectionGene((if (rand.nextBoolean()) geneMap1 else geneMap2)[key]!!)
                    if (geneMap1[key]!!.isEnabled != geneMap2[key]!!.isEnabled) trait.isEnabled =
                        rand.nextDouble() >= 0.75f
                } else if (parent1.fitness != parent2.fitness) trait = geneMap1[key]
                else {               // disjoint or excess and equal fitness
                    trait = if (geneMap1.containsKey(key)) geneMap1[key] else geneMap2[key]
                    if (rand.nextBoolean()) continue
                }
                trait?.let { child.connections.add(it) }
            }
            return child
        }

        @JvmStatic
        fun isSameSpecies(g1: Genome, g2: Genome): Boolean {
            val sz1 = g1.connections.size
            val sz2 = g2.connections.size
            val combinedSize = sz1 + sz2

            val geneMap1 = g1.connections.iterator()
            val geneMap2 = g2.connections.iterator()

            val matchSize = combinedSize / max(1, sim.SPECIES_DENOMINATOR)
            val rollingSize = min(sz1, sz2)
            var tally = 0
            if (rollingSize >= matchSize) {
                val mette = FastSet<Int>()//combinedSize - matchSize
                val intRange = 0 until max(sz1, sz2)
                for (x: Int in intRange) {
                    if (geneMap1.hasNext() && !mette.add(geneMap1.next().innovationKey)) tally++
                    if (geneMap2.hasNext() && !mette.add(geneMap2.next().innovationKey)) tally++
                    if (tally >= matchSize) return true
                }
            }
            return false
        }

        lateinit var sim: Sim
        val INDEXABLE: Int by lazy { (sim).run { INPUTS + HIDDEN_NODES } }

    }

    override fun toString(): String {
        val unqLinks = connections.filter { it.isEnabled }
        val inList: List<Int> = unqLinks.map { it.keyInto }.distinct()
        val outList: List<Int> = unqLinks.map { it.keyOut }.distinct()
        val allinks: List<Int> = inList + outList
        return "Genome(fitness=$fitness, points=$points, adjustedFitness=$adjustedFitness," +
                "mutationRates=$mutationRates, connectionGeneList={" +
                "size=${unqLinks.size}, in=$inList, out=$outList }}}, nodes={size=${nodes.size}," +
                "nodes=${
                    nodes.filtered {
                        it.key in allinks
                    }.map {
                        it.impulse to it.incomingCon.size to it.activationFunction
                    }
                }})"
    }

}

inline operator fun FastSortedTable<NodeGene>.unaryMinus() = { x: Int -> this.filtered { x == it.key }.first() }
val misses = AtomicInteger(0)
val accesses = AtomicInteger(0)
inline operator fun FastSortedTable<NodeGene>.unaryPlus() = { x: Int ->
    accesses.getAndIncrement().let {
        if (x > INDEXABLE) (-this)(x) else
            this[x].takeIf { sim.INPUTS >= x || x == it.key }
                ?: (-this)(x).also { misses.getAndIncrement() }
    }
}

